File: Design Pattern/Behavioral/Template Method/order_processing.cpp
#include <iostream>
#include <string>
using namespace std;

class Order {
  string id;
  double subtotal;

public:
  Order(string id, double subtotal) : id(id), subtotal(subtotal) {}
  string getId() { return id; }
  double getSubtotal() { return subtotal; }
};

class OrderProcessor {
public:
  virtual ~OrderProcessor() = default;

  void processOrder(Order *order) {
    validateOrder(order);
    calculateTotal(order);
    applyDiscount(order);
    processPayment(order);
    sendConfirmation(order);
    cout << "Order processing complete: " << order->getId() << endl;
  }

protected:
  virtual void validateOrder(Order *order) = 0;
  virtual void calculateTotal(Order *order) = 0;
  virtual void processPayment(Order *order) = 0;

  virtual void applyDiscount(Order *order) {}

  virtual void sendConfirmation(Order *order) {
    cout << "Sending email confirmation for order " << order->getId() << endl;
  }
};

class StandardOrderProcessor : public OrderProcessor {
protected:
  void validateOrder(Order *order) override {
    cout << "Validating standard order: checking stock and address..." << endl;
  }

  void calculateTotal(Order *order) override {
    double shipping = 5.99;
    double total = order->getSubtotal() + shipping;
    cout << "Standard total: $" << total << " (including $" << shipping
         << " shipping)" << endl;
  }

  void processPayment(Order *order) override {
    cout << "Processing payment via standard gateway..." << endl;
  }
};

class PrimeOrderProcessor : public OrderProcessor {
protected:
  void validateOrder(Order *order) override {
    cout << "Validating Prime order: checking membership and stock..." << endl;
  }

  void calculateTotal(Order *order) override {
    cout << "Prime total: $" << order->getSubtotal() << " (free shipping)"
         << endl;
  }

  void processPayment(Order *order) override {
    cout << "Processing payment via Prime billing..." << endl;
  }

  void applyDiscount(Order *order) override {
    cout << "Applying 10% Prime member discount..." << endl;
  }
};

class InternationalOrderProcessor : public OrderProcessor {
protected:
  void validateOrder(Order *order) override {
    cout << "Validating international order: customs, restricted items, "
            "address format..."
         << endl;
  }

  void calculateTotal(Order *order) override {
    double shipping = 24.99;
    double customs = order->getSubtotal() * 0.15;
    double total = order->getSubtotal() + shipping + customs;
    cout << "International total: $" << total << " (shipping: $" << shipping
         << ", customs: $" << customs << ")" << endl;
  }

  void processPayment(Order *order) override {
    cout << "Processing payment with currency conversion..." << endl;
  }

  void sendConfirmation(Order *order) override {
    cout << "Sending multi-language confirmation with tracking for order "
         << order->getId() << endl;
  }
};

int main() {
  Order order1("ORD-001", 49.99);
  Order order2("ORD-002", 149.99);
  Order order3("ORD-003", 89.99);

  StandardOrderProcessor standard;
  standard.processOrder(&order1);

  cout << endl;

  PrimeOrderProcessor prime;
  prime.processOrder(&order2);

  cout << endl;

  InternationalOrderProcessor international;
  international.processOrder(&order3);

  return 0;
}

File: Design Pattern/Behavioral/Template Method/order_processing.ts
class Order {
  public id: string;
  public subtotal: number;

  constructor(id: string, subtotal: number) {
    this.id = id;
    this.subtotal = subtotal;
  }
}

abstract class OrderProcessor {
  public processOrder(order: Order): void {
    this.validateOrder(order);
    this.calculateTotal(order);
    this.applyDiscount(order);
    this.processPayment(order);
    this.sendConfirmation(order);
    console.log("Order processing complete: " + order.id);
  }

  protected abstract validateOrder(order: Order): void;
  protected abstract calculateTotal(order: Order): void;
  protected abstract processPayment(order: Order): void;

  protected applyDiscount(order: Order): void {}

  protected sendConfirmation(order: Order): void {
    console.log("Sending email confirmation for order " + order.id);
  }
}

class StandardOrderProcessor extends OrderProcessor {
  protected validateOrder(order: Order): void {
    console.log("Validating standard order: checking stock and address...");
  }

  protected calculateTotal(order: Order): void {
    const shipping = 5.99;
    const total = order.subtotal + shipping;
    console.log(`Standard total: $${total} (including $${shipping} shipping)`);
  }

  protected processPayment(order: Order): void {
    console.log("Processing payment via standard gateway...");
  }
}

class PrimeOrderProcessor extends OrderProcessor {
  protected validateOrder(order: Order): void {
    console.log("Validating Prime order: checking membership and stock...");
  }

  protected calculateTotal(order: Order): void {
    console.log(`Prime total: $${order.subtotal} (free shipping)`);
  }

  protected processPayment(order: Order): void {
    console.log("Processing payment via Prime billing...");
  }

  protected applyDiscount(order: Order): void {
    console.log("Applying 10% Prime member discount...");
  }
}

class InternationalOrderProcessor extends OrderProcessor {
  protected validateOrder(order: Order): void {
    console.log(
      "Validating international order: customs, restricted items, address format..."
    );
  }

  protected calculateTotal(order: Order): void {
    const shipping = 24.99;
    const customs = order.subtotal * 0.15;
    const total = order.subtotal + shipping + customs;
    console.log(
      `International total: $${total} (shipping: $${shipping}, customs: $${customs})`
    );
  }

  protected processPayment(order: Order): void {
    console.log("Processing payment with currency conversion...");
  }

  protected sendConfirmation(order: Order): void {
    console.log(
      "Sending multi-language confirmation with tracking for order " + order.id
    );
  }
}

const order1 = new Order("ORD-001", 49.99);
const order2 = new Order("ORD-002", 149.99);
const order3 = new Order("ORD-003", 89.99);

const standard: OrderProcessor = new StandardOrderProcessor();
standard.processOrder(order1);

console.log();

const prime: OrderProcessor = new PrimeOrderProcessor();
prime.processOrder(order2);

console.log();

const international: OrderProcessor = new InternationalOrderProcessor();
international.processOrder(order3);
